### PR TITLE
Beacon chain Merkle tree APIs

### LINKED ISSUE #1673: Add `get_merkle_root`
### How can it be fixed
- Reorg `eth/_utils/merkle.py`
- Add `get_merkle_root`

### PR DESCRIPTION
**Pull request by hwwhww**
### What was wrong?
Fixes #1673

### How was it fixed?
1. Refactor `eth._utils.merkle` APIs
2. Add `get_merkle_root` and `get_updated_merkle_accumulator` (per https://github.com/ethereum/eth2.0-specs/pull/354)

TODO in other PRs:
Not a fan of the dependency of `eth._utils.merkle -> eth.beacon._utils.hash`. I'm thinking about moving them to `eth2._utils` or `eth2.common` if we do https://github.com/ethereum/py-evm/issues/1659 in other PRs.

#### Cute Animal Picture

xmas_turtle

### CODE REVIEW

**Comment by hwwhww** [comment]
@djrtwo 
Hmm, I *think* it's okay to merge now because (i) it's useful anyway even without ethereum/eth2.0-specs#354 and (ii) the new withdrawable PR might take more time to be merged.

File: eth/beacon/_utils/hash.py [hunk 1]
```diff
@@ -1,8 +1,12 @@
+from typing import (
+    Hashable,
+)
+
 from eth_typing import Hash32
 from eth_hash.auto import keccak
 
 
-def hash_eth2(data: bytes) -> Hash32:
+def hash_eth2(data: Hashable) -> Hash32:
```
Comment (ChihChengLiang) [inline eth/beacon/_utils/hash.py]: I might miss some historical discussion. But is `Hashable` what we want here?
According to typing documentation, `Hashable` is an alias to `collections.abc.Hashable`, which checks if an object has the abstract method `__hash__`. That is in the sense of Python `hash` rather then if a `Serializable` object could be hash_eth2-ed or tree-hashed.

File: eth/_utils/merkle.py [hunk 2]
```diff
@@ -64,16 +71,20 @@ def get_merkle_proof(tree: MerkleTree, item_index: int) -> Sequence[Hash32]:
 
 
 def _calc_parent_hash(left_node: Hash32, right_node: Hash32) -> Hash32:
-    """Calculate the parent hash of a node and its sibling."""
-    return keccak(left_node + right_node)
+    """
+    Calculate the parent hash of a node and its sibling.
+    """
+    return hash_eth2(left_node + right_node)
```
Comment (ChihChengLiang) [inline eth/_utils/merkle.py]: Are these APIs dedicated to eth2.0?

Comment (hwwhww) [reply eth/_utils/merkle.py]: Currently, yes. It was written for proof of custody: https://github.com/ethereum/py-evm/pull/919

**Pull request closed without merging**